tests/integration/conftest.py

```diff
@@ -154,3 +167,50 @@ def load(self):
 
     def save(self, tables: TablesCrawler, workspace_info: WorkspaceInfo) -> str:
         raise RuntimeWarning("not available")
+
+
+class StaticServicePrincipalMigration(ServicePrincipalMigration):
+    def save(self, migration_results: list[StorageCredentialValidationResult]) -> str:
+        return "azure_service_principal_migration_result.csv"
+
+
+class StaticStorageCredentialManager(StorageCredentialManager):
+    # During integration test, we only want to list storage_credentials that are created during the test.
+    # So we provide a credential name list so the test can ignore credentials that are not in the list.
+    def __init__(self, ws_client: WorkspaceClient, credential_names: set[str]):
+        super().__init__(ws_client)
+        self._credential_names = credential_names
+
+    def list_storage_credentials(self) -> set[str]:
+        application_ids = set()
+
+        storage_credentials = self._ws.storage_credentials.list(max_results=0)
+
+        for storage_credential in storage_credentials:
+            if not storage_credential.azure_service_principal:
+                continue
+            if storage_credential.name in self._credential_names:
+                application_ids.add(storage_credential.azure_service_principal.application_id)
+
+        logger.info(
+            f"Found {len(application_ids)} distinct service principals already used in storage credentials during integration test"
+        )
+        return application_ids
```

Comment:
```suggestion
```
use `create_autospec(WorkspaceClient)` and `ws.storage_credentials.list.return_value = [...]` in unit tests. For integration tests, just add `include_names: list[str] | None` argument to `ServicePrincipalMigration.run()` method.

Having `StorageCredentialManager` makes code more complicated than it should

Reply:
fixed